# Day 7 - The Treachery of Whales


Link to page: https://adventofcode.com/2021/day/7

A giant whale has decided your submarine is its next meal, and it's much faster than you are. There's nowhere to run!

Suddenly, a swarm of crabs (each in its own tiny submarine - it's too deep for them otherwise) zooms in to rescue you! They seem to be preparing to blast a hole in the ocean floor; sensors indicate a massive underground cave system just beyond where they're aiming!

The crab submarines all need to be aligned before they'll have enough power to blast a large enough hole for your submarine to get through. However, it doesn't look like they'll be aligned before the whale catches you! Maybe you can help?

There's one major catch - crab submarines can only move horizontally.

You quickly make a list of the horizontal position of each crab (your puzzle input). Crab submarines have limited fuel, so you need to find a way to make all of their horizontal positions match while requiring them to spend as little fuel as possible.

For example, consider the following horizontal positions:

16,1,2,0,4,2,7,1,2,14
This means there's a crab with horizontal position 16, a crab with horizontal position 1, and so on.

Each change of 1 step in horizontal position of a single crab costs 1 fuel. You could choose any horizontal position to align them all on, but the one that costs the least fuel is horizontal position 2:
<code>
Move from 16 to 2: 14 fuel
Move from 1 to 2: 1 fuel
Move from 2 to 2: 0 fuel
Move from 0 to 2: 2 fuel
Move from 4 to 2: 2 fuel
Move from 2 to 2: 0 fuel
Move from 7 to 2: 5 fuel
Move from 1 to 2: 1 fuel
Move from 2 to 2: 0 fuel
Move from 14 to 2: 12 fuel
</code>
This costs a total of 37 fuel. This is the cheapest possible outcome; more expensive outcomes include aligning at position 1 (41 fuel), position 3 (39 fuel), or position 10 (71 fuel).

In [11]:
import pandas as pd
import numpy as np
import sys
np.set_printoptions(threshold=sys.maxsize)
#np.set_printoptions(threshold=50)


In [12]:
pwd

'C:\\Users\\incardo\\PycharmProjects\\advent_of_code_2021'

In [77]:
with open('input_files/day7_input.txt') as f:
    lines = f.read().splitlines()

crab_list = np.array([int(x) for x in lines[0].split(",")])
crab_list

array([1101,    1,   29,   67, 1102,    0,    1,   65, 1008,   65,   35,
         66, 1005,   66,   28,    1,   67,   65,   20,    4,    0, 1001,
         65,    1,   65, 1106,    0,    8,   99,   35,   67,  101,   99,
        105,   32,  110,   39,  101,  115,  116,   32,  112,   97,  115,
         32,  117,  110,  101,   32,  105,  110,  116,   99,  111,  100,
        101,   32,  112,  114,  111,  103,  114,   97,  109,   10,   48,
        221,  311,  159,  144, 1257, 1247,   28,  567,  700,  110,  711,
        602, 1033, 1260, 1403,  573,   14,  214,  305,  229,  541,  165,
        314, 1924,  465,  806,  649,    1, 1220,  342,   80,  168,  288,
        222,   92,    4, 1239, 1319, 1145,  233,  557, 1285,  963,  291,
       1283,   55, 1617,   17,  379, 1373,  320,  300,  632, 1457,  166,
        282,    8,   36,   66,   32,  376,  156,  482,  291,  512,   62,
        172,   92, 1389,    4,  286,  168,  619, 1781,  785,  468,  300,
        820, 1070,  213,  304,   66,  106,   89,  1

### Part 1

Determine the horizontal position that the crabs can align to using the least fuel possible. How much fuel must they spend to align to that position?

In [84]:
def calc_fuel_alignment(crab_list):
    target = np.median(crab_list)
    fuel = np.sum(abs(crab_list-target))
    return fuel

calc_fuel_alignment(crab_list)

331067.0

### Part 2

The crabs don't seem interested in your proposed solution. Perhaps you misunderstand crab engineering?

As it turns out, crab submarine engines don't burn fuel at a constant rate. Instead, each change of 1 step in horizontal position costs 1 more unit of fuel than the last: the first step costs 1, the second step costs 2, the third step costs 3, and so on.

As each crab moves, moving further becomes more expensive. This changes the best horizontal position to align them all on; in the example above, this becomes 5:
<code>
Move from 16 to 5: 66 fuel
Move from 1 to 5: 10 fuel
Move from 2 to 5: 6 fuel
Move from 0 to 5: 15 fuel
Move from 4 to 5: 1 fuel
Move from 2 to 5: 6 fuel
Move from 7 to 5: 3 fuel
Move from 1 to 5: 10 fuel
Move from 2 to 5: 6 fuel
Move from 14 to 5: 45 fuel
</code>
This costs a total of 168 fuel. This is the new cheapest possible outcome; the old alignment position (2) now costs 206 fuel instead.

Determine the horizontal position that the crabs can align to using the least fuel possible so they can make you an escape route! How much fuel must they spend to align to that position?

In [90]:
part1, part2 = float("inf"), float("inf")
#calculate the difference between each value of the array and all the possible values in the range (min, max). Then take the sum of differences which is smaller
    
for target in range(min(crab_list), max(crab_list)): 
    diffs = [abs(i-target) for i in crab_list]
    sums = [(n*(n+1))/2 for n in diffs] 

    part1 = min(part1, sum(diffs)) #re-assign the min at each iteration
    part2 = min(part2, sum(sums))  #re-assign the min at each iteration

print(part1, int(part2))

331067 92881128
